File: Python/16.3Sum Closest.py
#在一串数字中找三个数字，使它们的和最接近目标数字，返回最接近的和
class Solution(object):
    def threeSum(self, nums,target):
        nums.sort()         #先排序
        closest , dis =  nums[0]+nums[1]+nums[2] , abs(nums[0]+nums[1]+nums[2]-target)
        for i in range(0, len(nums)):
            if (i > 0 and nums[i] == nums[i-1]):
                continue     #和前面一个数一样的话，跳过
            l, r = i + 1, len(nums) - 1
            while l < r:
                sum = nums[l] + nums[r] + nums[i]
                if sum == target:
                    return sum
                elif sum < target:
                    l = l + 1
                else:
                    r = r - 1
                if abs(target-sum)<dis:
                    dis = abs(target-sum)
                    closest = sum
        return closest

s = Solution()
print(s.threeSum([-1,2,1,-4],1))

File: C++/Array/16.3Sum Closest.cpp
/*
求数组中的三个数和最接近target。返回这三个数的和。
*/
/*
思路：借用3Sum的方法。多一层循环，每一层把重复的跳过。
*/
#include <iostream>
#include <vector>
#include <algorithm>
#include <windows.h>
#include <windef.h>
#include <climits>
#include <cmath>
using namespace std;

class Solution {
public:
	int threeSumClosest(vector<int>& nums, int target) {
		int length = nums.size(), i, j, k,result,dis = INT_MAX,tmp;
		sort(nums.begin(), nums.end());
		for (k = 0; k < length - 2; ++k)
		{
			int tar = target-nums[k];
			i = k + 1, j = length - 1;
			while (i < j)
			{
				if (nums[i] + nums[j] == tar)	return target;
				tmp = fabs(nums[i] + nums[j]-tar);
				if(tmp<dis)
				{
					dis = tmp;
					result = nums[i] + nums[j] + nums[k];
				}
				if (nums[i] + nums[j] < tar)
				{
					++i;
					while (nums[i - 1] == nums[i]) ++i;
				}
				else
				{
					--j;
					while (j < length - 1 && nums[j + 1] == nums[j]) --j;
				}
			}
			while (nums[k] == nums[k + 1]) ++k;
		}
		return result;
	}

};

int main()
{
	Solution s;
	vector<int> nums{-1,2,1,-4};
	cout<<s.threeSumClosest(nums,1)<<endl;
}

